File: kotlin/math/Collatz.kt
// Print the start below 1000 with the longest Collatz sequence, picked by maxByOrNull over a range.
// Run: kotlinc Collatz.kt -include-runtime -d collatz.jar && java -jar collatz.jar

fun chainLength(start: Int): Int {
    var value = start.toLong()
    var length = 1
    while (value != 1L) {
        value = if (value % 2 == 0L) value / 2 else value * 3 + 1
        length++
    }
    return length
}

fun main() {
    val limit = 1000
    val longest = (1 until limit).maxByOrNull { chainLength(it) }!!
    println("$longest ${chainLength(longest)}")
}

File: julia/math/collatz.jl
# Print the start below 1000 with the longest Collatz sequence, picked by argmax over a range.
# Run: julia collatz.jl

const LIMIT = 1000

function chain_length(start)
    value = start
    length = 1
    while value != 1
        value = iseven(value) ? value ÷ 2 : value * 3 + 1
        length += 1
    end
    return length
end

starts = 1:(LIMIT - 1)
longest = starts[argmax(chain_length.(starts))]
println(longest, " ", chain_length(longest))
